libs/patchwheel/patch.go

```diff
@@ -0,0 +1,270 @@
+package patchwheel
+
+import (
+	"archive/zip"
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/databricks/cli/libs/log"
+)
+
+const (
+	versionKey = "Version:"
+	nameKey    = "Name:"
+)
+
+func findFiles(r *zip.ReadCloser, files ...string) []*zip.File {
+	found := 0
+	result := make([]*zip.File, len(files))
+	for _, f := range r.File {
+		for ind, prev := range result {
+			if prev != nil {
+				continue
+			}
+			if f.Name == files[ind] {
+				result[ind] = f
+				found += 1
+				if found >= len(files) {
+					return result
+				}
+			}
+		}
+	}
+	return result
+}
+
+// patchMetadata returns new METADATA content with an updated "Version:" field and validates that previous version matches oldVersion
+func patchMetadata(r io.Reader, oldVersion, newVersion string) ([]byte, error) {
+	scanner := bufio.NewScanner(r)
+	var buf bytes.Buffer
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, versionKey) {
+			foundVersion := strings.TrimSpace(line[len(versionKey):])
+			if foundVersion != oldVersion {
+				return nil, fmt.Errorf("Unexpected version in METADATA: %s (expected %s)", strings.TrimSpace(line), oldVersion)
+			}
+			line = versionKey + newVersion
+		}
+		buf.WriteString(line + "\n")
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// patchRecord updates RECORD content: it replaces the old dist-info prefix with the new one
+// in all file paths and, for the METADATA entry, updates the hash and size.
+func patchRecord(r io.Reader, oldDistInfoPrefix, newDistInfoPrefix, metadataHash string, metadataSize int) ([]byte, error) {
+	scanner := bufio.NewScanner(r)
+	var newLines []string
```

Comment:
This function accumulates lines while the previous function writes to a buffer. Can both use the same approach?

Reply:
It looks like it can write to a buffer directly as well.

Reply:
Fixed.